ships/0024-surfacing-errors-to-buildrun.md

```diff
@@ -117,11 +117,56 @@ As a third-party application developer that uses shipwright under the hood, I wa
+The new file `pkg/reconciler/buildrun/resources/failures.go` includes the extraced logic and other failure related functions.
+
+A TaskRun can emit failures using two results:
+1. shp-error-reason
+2. shp-error-message
+
+Whenever a Tekton TaskRun fails, it keeps a JSON string of all results inside each terminated step in `TaskRun.Status.Steps[i].Terminated.Message` of type `PipelineResourceResult`. Whenever we can find the two error results in a step's message field, we surface the error as `failureDetails`.
+
+So far, we have presented a mechanism to lift errors results into a BuildRun. However, no component has produced error results yet.  One source of failure is misconfiguration in the source step for git applications. For example, a user can forget to provide authentication details for a private repository. Therefore, we propose to implement error results for `cmd/git`. On top of that, we propose implementing smarter error reporting by parsing and classifying git's messages for errors in `stdout`. A third-party application would be able to provide mitigations for failed builds based on error classes in git, for example, a UI workflow to reenter basic authentication details.
+
+The following errors classes are implemented for git:
+- `AuthInvalidUserOrPass`, expresses that basic authentication is not possible
+- `AuthInvalidKey`, expresses that ssh authentication is not possible
+- `BranchNotFound` expresses that a remote branch does not exist
+- `RepositoryNotFound`, expresses that the remote target for the git operation does not exist. It triggers when an error message is enough to determine that the remote target does not exist and is mostly derived from the Git server's messages e.g. GitLab or GitHub
+- `AuthPrompted`, is caused when a repo is not found, is private and authentication is insufficient
+- `Unknown`, is the class of choice if no other class fits
+
+
+Each class has a helpful, static message to hint a user to a right direction to fix his build. The unknown class is generic and chosen if no other class matches. Its message is an exception since we can not determine the exact class. Thus, we propagate the git's stdout and publish it as a message.
+To prevent too-long messages, the failure message offers a setter function that cuts long error messages. Therefore, we stay within Tekton's limit for emitting results.
 
 ### Test Plan
 
-`tbd in the next phase of the proposal`
+To make sure that the new feature functions correctly, we propose a testing strategy consisting of e2e and unit tests.
+
+#### BuildRun and surfacing errors
+
+E2E tests are tricky for this feature since the error reporting is not part of it. We could implement a custom strategy for testing that implements error reporting (currently not planned). Alternatively, we can implement a proper e2e test by running a misconfigured source step. The e2e test would hit the surfacing feature and the new error reporting in our git wrapper.
```

Comment:
This IMO makes the most sense for end to end testing - configure a BuildRun that will fail on purpose.

Reply:
I will setup and end-to-end test that fails on purpose in the source step.